stripecardscan/src/main/java/com/stripe/android/stripecardscan/scanui/ScanFragment.kt

```diff
@@ -0,0 +1,277 @@
+                Manifest.permission.CAMERA,
+            ) == PackageManager.PERMISSION_GRANTED -> {
+                launch { permissionStat.trackResult("success") }
+                prepareCamera { onCameraReady() }
+            }
+            storage.getBoolean(
+                PERMISSION_RATIONALE_SHOWN,
+                false,
+            ) -> showPermissionDenied()
+            else -> requestCameraPermission()
+        }
+    }
+
+    /**
+     * Show text explaining that the camera is not available.
+     */
+    protected open fun showCameraNotSupported() {
+        instructionsText.visibility = View.VISIBLE
+        instructionsText.setText(R.string.stripe_error_camera_unsupported)
+        scanFailure()
+    }
+
+    /**
+     * Show text  for why we are requesting camera permissions when the permission
+     * has been denied.
+     */
+    protected open fun showPermissionDenied() {
+        instructionsText.visibility = View.VISIBLE
+        instructionsText.setText(R.string.stripe_camera_permission_settings_message)
+        storage.storeValue(PERMISSION_RATIONALE_SHOWN, true)
+    }
+
+    /**
+     * Request permission to use the camera.
+     */
+    protected open fun requestCameraPermission() {
+        cameraPermissionLauncher.launch(
+            Manifest.permission.CAMERA
+        )
+    }
+
+    /**
+     * Turn the flashlight on or off.
+     */
+    protected open fun toggleFlashlight() {
+        isFlashlightOn = !isFlashlightOn
+        setFlashlightState(isFlashlightOn)
+        // TODO: this should be reported as part of scanstats, but is not yet supported
+    }
+
+    /**
+     * Toggle between available cameras.
+     */
+    protected open fun toggleCamera() {
+        cameraAdapter.changeCamera()
+        // TODO: this should probably be reported as part of scanstats, but is not yet supported
+    }
+
+    /**
+     * Called when the flashlight state has changed.
+     */
+    protected abstract fun onFlashlightStateChanged(flashlightOn: Boolean)
+
+    /**
+     * Turn the flashlight on or off.
+     */
+    private fun setFlashlightState(on: Boolean) {
+        cameraAdapter.setTorchState(on)
+        isFlashlightOn = on
+        onFlashlightStateChanged(on)
+    }
+
+    /**
+     * Cancel scanning due to a camera error.
+     */
+    protected open fun scanFailure(cause: Throwable? = null) {
+        Log.e(Config.logTag, "Canceling scan due to error", cause)
+        runBlocking { scanStat.trackResult("scan_failure") }
+        resultListener.failed(cause)
+        closeScanner()
+    }
+
+    /**
+     * The scan has been closed by the user.
+     */
+    protected open fun userClosedScanner() {
+        runBlocking { scanStat.trackResult("user_canceled") }
+        resultListener.userCanceled(CancellationReason.Closed)
+        closeScanner()
+    }
+
+    /**
+     * Close the scanner.
+     */
+    protected open fun closeScanner() {
+        setFlashlightState(false)
```

Comment:
Can we stop the scanning process here? You should be able to cancel the scan flow to halt processing.